### ISSUE TITLE
Flutter driver does not work with internationalization

### ISSUE THREAD

**Issue by mmcc007** [flutter/flutter#27785]
I am testing an internationalized version of the default counter app using the default (unmodified) counter integration test. 

The internationalized default app runs fine if run 'live' (using 'flutter run') on an emulator/simulator before and after the locale is changed (tried changing locale via settings and via the command line).
<details>
    <summary>screenshots</summary>

![screenshot_1549912943](https://user-images.githubusercontent.com/1980785/52588137-a844e880-2df0-11e9-935a-04fbb86a6a39.png)
![simulator screen shot - iphone-x - 2019-02-11 at 11 28 55](https://user-images.githubusercontent.com/1980785/52588138-a8dd7f00-2df0-11e9-89d0-94663846a967.png)

</details>

However, when I attempt to run the default integration test (... via flutter driver) on the internationalized default app, the test passes for the default locale and fails when the locale is changed to the new locale. This result is consistent if the integration test is run on an emulator or a simulator (have not yet tested on android/ios devices). This indicates there is a flutter driver problem related to the flutter implementation of internationalization (unless there is a bug in the code implementing internationalization.. code is available at https://github.com/mmcc007/test_internationalization/tree/master/default_app_int).

I have an online sample of this running on travis - output https://travis-ci.org/mmcc007/test_internationalization/builds/491802376 running on master channel. 

As a sanity test, the sample, using flutter driver, first runs the default integration test on the default app ([unmodified](https://github.com/mmcc007/test_internationalization/tree/d5fae044bf7a7b1d119433b37cfd830fa9b9c1ff/default_app)) on an iOS simulator set to the locale 'english' (the default locale) and then to the locale 'french'. The default integration test runs fine for the default app in either locale (as would be expected). 

Then the sample runs the internationalized version of the app. The same default (unmodified) integration test succeeds for the locale ('english') and [fails](https://travis-ci.org/mmcc007/test_internationalization/jobs/491802378) for the locale 'french'.

You can see the error output https://travis-ci.org/mmcc007/test_internationalization/jobs/491802378#L525 and below:

```
flutter: Observatory listening on http://127.0.0.1:51101/
00:00 +0: end-to-end test (setUpAll)
[info ] FlutterDriver: Connecting to Flutter application at http://127.0.0.1:51101/
[trace] FlutterDriver: Isolate found with number: 626846213
[trace] FlutterDriver: Isolate is paused at start.
[trace] FlutterDriver: Attempting to resume isolate
[trace] FlutterDriver: Waiting for service extension
[info ] FlutterDriver: Connected to Flutter application.
00:02 +0: end-to-end test tap on the floating action button; verify counter
[warning] FlutterDriver: waitFor message is taking a long time to complete...
00:32 +0 -1: end-to-end test tap on the floating action button; verify counter [E]
  TimeoutException after 0:00:30.000000: Test timed out after 30 seconds.
```

<details>
    <summary>The 'flutter doctor -v' for the failing test is:</summary>

```
[✓] Flutter (Channel master, v1.2.1-pre.144, on Mac OS X 10.13.3 17D102, locale en-US)
    • Flutter version 1.2.1-pre.144 at /Users/travis/build/mmcc007/test_internationalization/flutter
    • Framework revision 20d4770136 (8 minutes ago), 2019-02-11 14:39:28 -0500
    • Engine revision 5dee035ee2
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 174d6fec3d)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
      If Android SDK has been installed to a custom location, set ANDROID_HOME to that location.
      You may also want to add it to your PATH environment variable.
[!] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode-9.4.1.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    ✗ Verify that all connected devices have been paired with this computer in Xcode.
      If all devices have been paired, libimobiledevice and ideviceinstaller may require updating.
      To update with Brew, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew uninstall --ignore-dependencies usbmuxd
        brew install --HEAD usbmuxd
        brew unlink usbmuxd
        brew link usbmuxd
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    • ios-deploy 1.9.2
    ✗ ios-deploy out of date (1.9.4 is required). To upgrade with Brew:
        brew upgrade ios-deploy
    • CocoaPods version 1.5.3
[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).
[!] Connected device
    ! No devices available
! Doctor found issues in 4 categories.
```

</details>
The same result can be found for the stable channel https://travis-ci.org/mmcc007/test_internationalization/jobs/491485096.

**Comment by ykorzikowski** [comment]
what is the status here? Is there any way to set locales with flutter drive?

**Comment by pakivika** [comment]
Oh, I worked hard to get screenshots automated for several languages, and now encountered this bug. What a waste of time :-(

**Comment by mmcc007** [comment]
@pakivika Sorry to hear about the roadblock to getting your screenshots working.

While this issue is open, some work is progressing on finding a workaround for this issue.
See mmcc007/screenshots/issues/19 for details

In the meantime, give the first comment (https://github.com/flutter/flutter/issues/27785#issue-408955077) a thumbs-up to prioritize this issue with the Flutter Team

**Comment by pakivika** [comment]
> Oh, I worked hard to get screenshots automated for several languages, and now encountered this bug. What a waste of time :-(

Have to revert back a bit with my words. It was not total waste of time, I can still trigger automated screenshots manually for each language (just have to kill simulator between languages). It's still saving me quite amount of time even when I cannot have CI taking care of screenshots.

**Comment by mmcc007** [comment]
In _'screenshots'_  the emulator and simulator are currently being restarted in new locale as follows:
https://github.com/mmcc007/screenshots/blob/13d973db2c6d2507bdea5f322bfc29d54ccffbbb/lib/screenshots.dart#L202-L207
https://github.com/mmcc007/screenshots/blob/13d973db2c6d2507bdea5f322bfc29d54ccffbbb/lib/screenshots.dart#L161-L170
... but, while the locale is changed on the 'device', 'flutter drive' is not picking it up. There is some kind of disconnect happening with 'flutter drive'.

Anyway, while this is definitely a bug, more experimentation might find a workaround.

**Comment by mmcc007** [comment]
FYI: just triggered another build of the demo project
https://github.com/mmcc007/test_internationalization
on latest flutter master showing the stacktrace for this issue on iOS. 
https://travis-ci.org/mmcc007/test_internationalization/jobs/552247331#L587

The stacktrace is the same for both iOS and Android:
```
 DriverError: Failed to fulfill WaitFor due to remote error
  Original error: Bad state: The client closed with pending request "ext.flutter.driver".
  Original stack trace:
  #0      new Client.withoutJson.<anonymous closure> (package:json_rpc_2/src/client.dart:70:24)
  #1      StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
  #2      StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:119:48)
  #3      _rootRun (dart:async/zone.dart:1120:38)
  #4      _CustomZone.run (dart:async/zone.dart:1021:19)
  #5      _FutureListener.handleWhenComplete (dart:async/future_impl.dart:150:18)
  #6      Future._propagateToListeners.handleWhenCompleteCallback (dart:async/future_impl.dart:609:39)
  #7      Future._propagateToListeners (dart:async/future_impl.dart:665:37)
  #8      Future._propagateToListeners (dart:async/future_impl.dart:566:9)
  #9      Future._completeWithValue (dart:async/future_impl.dart:483:5)
  #10     Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:513:7)
  #11     StackZoneSpecification._run (package:stack_trace/src/stack_zone_specification.dart:209:15)
  #12     StackZoneSpecification._registerCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:119:48)
  #13     _rootRun (dart:async/zone.dart:1124:13)
  #14     _CustomZone.run (dart:async/zone.dart:1021:19)
  #15     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
  #16     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
  #17     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
  #18     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
  #19     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
  #20     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
  #21     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

**Comment by mehmetf** [comment]
Just curious...

Has anyone tried to circumvent this by registering a handler for locale changes via https://api.flutter.dev/flutter/flutter_driver_extension/enableFlutterDriverExtension.html?

You can send the new locale information via https://api.flutter.dev/flutter/flutter_driver/FlutterDriver/requestData.html.

This is not the same thing as ondevice testing of course since you are not changing the locale of the device itself but it sounds like it should work to grab, say, screenshots in a certain locale.

**Comment by mmcc007** [comment]
@mehmetf looked into your suggestion, and in the process noticed the reason flutter driver was throwing a stacktrace. The test was using a tooltip to find a button whose text had been changed during  internationalization. So the test couldn't find the button and was hanging and eventually failing.

Using your approach I was able to fetch the text from the app for the current locale and use it to find the button. The test then ran successfully.
https://github.com/mmcc007/screenshots/blob/%2319_%2320_localization/example/test_driver/main.dart#L11-L25 (temporary branch)

Here is a successful run of a test in multiple locales on multiple devices (locale is being changed before each instance of the test run on each device):
https://travis-ci.com/mmcc007/screenshots/builds/119053191

So looks like I will be closing this issue soon (after a little more testing). 

The stacktrace did not give any hints to the pretty obvious (in retrospect) problem. Perhaps an issue should be filed for improving the flutter driver stacktrace when a test fails due to a problem with the test.
